mindsdb/integrations/handlers/lancedb_handler/lancedb_handler.py

```diff
@@ -0,0 +1,372 @@
+        offset: int = None,
+        limit: int = None,
+    ) -> HandlerResponse:
+        try:
+            # Load collection table
+            collection = self._client.open_table(table_name)
+        except Exception as e:
+            return Response(
+                resp_type=RESPONSE_TYPE.ERROR,
+                error_message=f"Error loading collection {table_name}: {e}",
+            )
+        filters = self._translate_condition(conditions)
+        # check if embedding vector filter is present
+        vector_filter = (
+            []
+            if conditions is None
+            else [
+                condition
+                for condition in conditions
+                if condition.column == TableField.SEARCH_VECTOR.value
+            ]
+        )
+
+        if len(vector_filter) > 0:
+            vector_filter = vector_filter[0]
+        else:
+            vector_filter = None
+
+        if vector_filter is not None:
+            vec = json.loads(vector_filter.value) if isinstance(vector_filter.value, str) else vector_filter.value
+            result = collection.search(vec).select(columns).to_pandas()
+            result = result.rename(columns={"_distance": TableField.DISTANCE.value})
+        else:
+            result = self._client.open_table(table_name).to_pandas()
+
+        new_columns = columns + [TableField.DISTANCE.value] if TableField.DISTANCE.value in result.columns else columns
+
+        col_str = ', '.join([col for col in new_columns if col in (TableField.ID.value, TableField.CONTENT.value, TableField.METADATA.value, TableField.EMBEDDINGS.value, TableField.DISTANCE.value)])
+
+        where_str = f'where {filters}' if filters else ''
+        if limit and offset:
+            sql = f"""select {col_str} from result {where_str} limit {limit} offset {offset}"""
+        elif limit and not offset:
+            sql = f"""select {col_str} from result {where_str} limit {limit}"""
+        elif offset and not limit:
+            sql = f"""select {col_str} from result {where_str} offset {offset}"""
+        else:
+            sql = f"""select {col_str} from result {where_str}"""
+
+        data_df = duckdb.query(sql).to_df()
+        return Response(resp_type=RESPONSE_TYPE.TABLE, data_frame=data_df)
+
+    def insert(
+        self, table_name: str, data: pd.DataFrame, columns: List[str] = None
+    ) -> HandlerResponse:
+        """
+        Insert data into the LanceDB database.
+        """
+        try:
+            collection = self._client.open_table(table_name)
+            df = data[[TableField.ID.value, TableField.CONTENT.value, TableField.METADATA.value, TableField.EMBEDDINGS.value]]
+            pa_data = pa.Table.from_pandas(df, preserve_index=False)
+            vec_data = vec_to_table(df[TableField.EMBEDDINGS.value].values.tolist())
+            new_pa_data = pa_data.append_column("vector", vec_data["vector"])
+            collection.add(new_pa_data)
+        except pa.lib.ArrowNotImplementedError:
```

Comment:
This looks confusing. What is happening here? Can you also add detailed comments to the insert method?

Reply:
Hi @ZoranPandovski , when a pandas dataframe is sent to a lancedb dataset which is empty (0 rows), the column type does not usually match and the exception pa.lib.ArrowNotImplementedError is thrown, in that case the dataset is recreated (only when the initial dataset is empty). Let me know if you have any other idea in mind.